Build docker images when building packages

jsoriano opened the pull request:
Include generation of docker images in the process of building the rest of released packages by extending `mage package`. Ideally it could be extended to build images for windows too.

Dockerfile, entry point templates and default beats configurations based on [current ones](https://github.com/elastic/beats-docker/tree/master/templates).

Pending:
- [x] Fix permission issues
- [x] Export built image as artifact with `docker save` or similar
- [x] Complete parametrization of docker options
- [x] Make it work for all beats
  - [x] Support capabilities for packetbeat and heartbeat
  - [x] Run as root for auditbeat
  - [x] Heartbeat user is `heartbeat-elastic` :thinking: it should be `heartbeat` as other beats
  - [x] ~~Install libsystemd in journalbeat~~ (systemd-libs already installed in centos:7 base image)
  - [x] Support community beats
- [x] Check that beats can be configured using environment variables
- [x] Fix license label for licensed builds
- [x] Add packaging tests for docker

andrewkroh commented on dev-tools/packaging/templates/docker/Dockerfile.tmpl:
```diff
@@ -0,0 +1,40 @@
+
+LABEL \
+  org.label-schema.schema-version="1.0" \
+  org.label-schema.vendor="Elastic" \
```
Community beats may want to reuse of image builders so this should be templated.

jsoriano replied:
Parameterized now.

andrewkroh commented on dev-tools/packaging/templates/docker/Dockerfile.tmpl:
```diff
@@ -0,0 +1,46 @@
+  org.label-schema.name="{{ .BeatName }}" \
+  org.label-schema.version="{{ beat_version }}" \
+  org.label-schema.url="{{ .BeatURL }}" \
+  org.label-schema.vcs-url="{{ $repoInfo.RootImportPath }}" \
```
I think adding org.label-schema.vcs-ref would be nice as well.

jsoriano replied:
Added.

jsoriano commented:
@andrewkroh @graphaelli some sanity tests are still pending, but I think this is ready for a first review, it'd be nice if you could give it another look. Thanks!

graphaelli commented on dev-tools/packaging/packages.yml:
```diff
@@ -157,6 +157,19 @@ shared:
+    extra_vars:
+      from: 'centos:7'
+      user: '{{ .BeatName }}'
+      repository: 'docker.elastic.co/beats'
```
can we add a param for this?

jsoriano replied:
As this is specific of docker packaging I left it as an extra var (like the specific extra vars for DMG builds). This extra var is optional, and can be set from `packages.yml` or overriden with `spec.ExtraVar()` from the magefile.

I am open to move it to other kind of parametrization though.

jsoriano replied:
I have added 8e6a32356 so it is not set by default on community beats.

graphaelli commented on dev-tools/mage/dockerbuilder.go:
```diff
@@ -0,0 +1,143 @@
+		}
+		outputFile = filepath.Join(distributionsDir, outputTar)
+	}
+	if err := sh.Run("docker", "save", "-o", outputFile, tag); err != nil {
```
I like this but it doesn't appear to be required for releases - the publisher currently does this + compression

jsoriano replied:
Oh, I though we also needed to generate the artifact and publisher does `docker save` and `docker push`.
I am using it also for sanity checks, so I'd leave it if it is not a problem.

graphaelli replied:
Not a problem from my perspective, agreed on keeping it.

graphaelli commented on dev-tools/packaging/packages.yml:
```diff
@@ -157,6 +157,19 @@ shared:
+      from: 'centos:7'
+      user: '{{ .BeatName }}'
+      repository: 'docker.elastic.co/beats'
+      linux_capabilities: ''
```
feature request: something like `expose_port` - and then use like [this](https://github.com/elastic/apm-server-docker/blob/b1b58e119edd106eb2882fb5e3db6e2b9d9ef734/templates/Dockerfile.j2#L37)

jsoriano replied:
Done, `expose_ports` can be a comma separated list of ports.

graphaelli commented on dev-tools/mage/dockerbuilder.go:
```diff
@@ -0,0 +1,143 @@
+	"strings"
+
+	"github.com/magefile/mage/sh"
+
```
new `goimports` doesn't like this space

jsoriano commented on .gitignore:
```diff
@@ -8,7 +8,7 @@
 /*/*.template*.json
 **/html_docs
-/*/_meta/kibana.generated
+/**/_meta/kibana.generated
```
Ok, I'll remove the change from here.

jsoriano commented on dev-tools/packaging/package_test.go:
```diff
@@ -311,6 +332,30 @@ func checkModules(t *testing.T, name, prefix string, r *regexp.Regexp, p *packag
+				t.Fatalf("%s entrypoint mode is %s, expected: %s", entrypoint, mode, expectedMode)
+			}
+		} else {
+			t.Skip("TODO: check if binary is in $PATH")
```
Ok, I'll change it to `t.Fatal`.

jsoriano commented:
@andrewkroh thanks for the review, I have addressed your comments.

Regarding additional tests, it could be possible, we could maybe modify system tests to use docker, so the docker image is also tested with them, but I don't have any plan for it at the moment.
